pkg/daemon/update.go

```diff
@@ -556,6 +613,134 @@ func (dn *Daemon) updateKernelArguments(oldConfig, newConfig *mcfgv1.MachineConf
 	return exec.Command("rpm-ostree", args...).Run()
 }
 
+// MountOSContainer mounts the container and returns the mountpoint
+func (dn *Daemon) MountOSContainer(container string) (mnt, containerName string, err error) {
+	var authArgs []string
+	if _, err = os.Stat(kubeletAuthFile); err == nil {
+		authArgs = append(authArgs, "--authfile", kubeletAuthFile)
+	}
+	// Pull the image
+	args := []string{"pull", "-q"}
+	args = append(args, authArgs...)
+	args = append(args, container)
+	pivotutils.RunExt(false, numRetriesNetCommands, "podman", args...)
+
+	containerName = "mcd-" + string(uuid.NewUUID())
+	// `podman mount` wants a container, so let's create a dummy one, but not run it
+	var cidBuf []byte
+	cidBuf, err = runGetOut("podman", "create", "--net=none", "--annotation=org.openshift.machineconfigoperator.pivot=true", "--name", containerName, container)
+	if err != nil {
+		return
+	}
+
+	cid := strings.TrimSpace(string(cidBuf))
+	// Use the container ID to find its mount point
+	mntBuf, err := runGetOut("podman", "mount", cid)
+	if err != nil {
+		return
+	}
+	mnt = strings.TrimSpace(string(mntBuf))
+	return
+}
+
+// switchKernel updates kernel on host with the kernelType specified in MachineConfig.
+// Right now it supports default (traditional) and realtime kernel
+func (dn *Daemon) switchKernel(oldConfig, newConfig *mcfgv1.MachineConfig) error {
+
+	// Do nothing if both old and new KernelType are of type default
+	if canonicalizeKernelType(oldConfig.Spec.KernelType) == defaultKernelType && canonicalizeKernelType(newConfig.Spec.KernelType) == defaultKernelType {
```

Comment:
Should we do nothing in general if they don’t differ?

Reply:
For kernelType=default (traditional kernel ) we don't need to do anything if kernelType in both old and new MC are same. I am using canonicalizeKernelType(), just to make sure that kernelType "" and default are treated same.

This is not the case when both new and old MC has kerneType=realtime, because in this case it could be possible that we have a new kernel-rt package update in latest OSContainer which we fetch from newconfig.spec.OSImageURL. If we don't do that kernel-rt package will never be updated on required nodes. For default kernel, since it is part of OSTree commit, it gets updated the usual way.

Does that make sense? or perhaps I didn't understand your question

Reply:
it does perfect cc @cgwalters